src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,362 @@
+                var originalNodes = new List<SyntaxNode>();
+                if (startTag != null)
+                {
+                    originalNodes.Add(startTag);
+                }
+                originalNodes.AddRange(tagChildren);
+                if (endTag != null)
+                {
+                    originalNodes.Add(endTag);
+                }
+
+                if (originalNodes.Count == 0)
+                {
+                    // Nothing to replace
+                    return parent;
+                }
+
+                // Replace nodes
+                var rewrittenElement = parent.ReplaceNodes(originalNodes, (original, rewritten) => {
+                    if (original == originalNodes[0] || rewritten == originalNodes[0])
+                    {
+                        // Put the new element in place of the first original node to be replaced.
+                        return markupElement;
+                    }
+
+                    return null;
+                });
+
+                return rewrittenElement;
+            }
+
+            private bool TryRecoverStartTag(
+                SyntaxNode parent,
+                string tagName,
+                MarkupTagBlockSyntax endTag,
+                int endTagIndex,
+                out SyntaxNode rewritten,
+                out int startTagIndex)
+            {
+                rewritten = parent;
+                startTagIndex = -1;
+                var malformedTagCount = 0;
+
+                foreach (var tracker in _startTagTracker)
+                {
+                    if (tracker.TagName.Equals(tagName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        break;
+                    }
+
+                    malformedTagCount++;
+                }
+
+                if (malformedTagCount != _startTagTracker.Count)
+                {
+                    parent = RewriteMalformedTags(parent, malformedTagCount);
+
+                    // One final rewrite, this is the rewrite that completes our target tag which is not malformed.
+                    var children = parent.ChildNodes();
+
+                    var startTagTracker = _startTagTracker.Pop();
+                    startTagIndex = startTagTracker.Index;
+                    var startTag = startTagTracker.TagBlock;
+                    
+                    // Get the nodes between the start and the end tag.
+                    var tagChildren = children.Skip(startTagIndex + 1).Take(endTagIndex - startTagIndex - 1).Cast<RazorSyntaxNode>();
+
+                    rewritten = RewriteNodeCore(parent, startTag, tagChildren, endTag);
+
+                    // We were able to recover
+                    return true;
+                }
+
+                // Could not recover tag. Aka we found an end tag without a corresponding start tag.
+                return false;
+            }
+
+            private SyntaxNode RewriteMalformedTags(SyntaxNode parent, int malformedTagCount)
```

Comment:
This method is invoked when `malformedTagCount != _startTagTracker.Count`, but reading the code it looks like `_startTagTracker.Count must be > malformedTagCount` (since it's popped that many times). Does the `if` need to be changed?

Reply:
Sure. It doesn't functionally change anything but I'll change the if.

Reply:
Done